### PR TITLE
[IA-3832] Resolve some circular imports in ajax/GoogleStorage

### PR DESCRIPTION
**Pull request by nawatts**
Split from #3447.

libs/ajax.js imports from libs/ajax/GoogleStorage.ts, which imports from libs/ajax.js.

This resolves that circular imports by...
- moving `withRequesterPays`, `fetchBuckets`, and `saToken` from ajax.js to ajax/GoogleStorage.ts
- moving `canUseWorkspaceProject` and the Billing request that it depends on to ajax/Billing.js
- Copy `getUser` from libs/auth.js to libs/state.js to allow getting the current user without importing from libs/auth (which imports from libs/ajax)

Interestingly, there's still a circular import in ajax-common: it imports from auth which imports from ajax, which imports from ajax-common. But removing that import from auth makes the tests fail, because of another circular import between libs/auth and libs/error.

This also moves the `checkBucketAccess` and `checkBucketLocation` requests from ajax.js to ajax/GoogleStorage.ts.

Reviewing commit by commit may be easier.

Co-authored-by: jdcanas@users.noreply.github.com

### CODE REVIEW

File: src/libs/utils.js [hunk 1]
```diff
@@ -89,9 +89,9 @@ export const hasAccessLevel = _.curry((required, current) => {
   return workspaceAccessLevels.indexOf(current) >= workspaceAccessLevels.indexOf(required)
 })
 
-export const canWrite = hasAccessLevel('WRITER')
-export const canRead = hasAccessLevel('READER')
-export const isOwner = hasAccessLevel('OWNER')
+export const canWrite = accessLevel => hasAccessLevel('WRITER', accessLevel)
+export const canRead = accessLevel => hasAccessLevel('READER', accessLevel)
+export const isOwner = accessLevel => hasAccessLevel('OWNER', accessLevel)
```
Comment (nawatts) [inline src/libs/utils.js:94]: Since `hasAccessLevel` is a curried function in a JS file, TS couldn't figure out that `canWrite`, etc. were callable with the way these were written.

File: src/libs/state.js [hunk 2]
```diff
@@ -17,6 +17,8 @@ export const authStore = Utils.atom({
   oidcConfig: {}
 })
 
+export const getUser = () => authStore.get().user
```
Comment (nawatts) [inline src/libs/state.js:20]: Copied from libs/auth.js to allow other modules to use this without pulling in all of auth's dependencies.

File: src/libs/ajax/Billing.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+  listAzureManagedApplications: async (subscriptionId, includeAssignedApplications) => {
+    const response = await fetchBillingProfileManager(
+      `azure/v1/managedApps?azureSubscriptionId=${subscriptionId}&includeAssignedApplications=${includeAssignedApplications}`,
+      _.merge(authOpts(), { signal }))
+    return response.json()
+  }
+})
+
+//TODO: this is a weird util method that calls an api... should be refactored so its not a circular utility
```
Comment (jdcanas) [inline src/libs/ajax/Billing.js]: This TODO can be removed now I believe. Might be worth wrapping it in the Billing module though, so it's clear it makes an API call.

File: src/libs/ajax/GoogleStorage.ts [hunk 4]
```diff
@@ -1,12 +1,77 @@
...
+    _.mergeAll([authOpts(token), jsonBody(scopes), { method: 'POST' }])
+  )
+  return res.json()
+}, {
+  expires: 1000 * 60 * 30,
+  // @ts-expect-error
+  keyFn: (...args) => JSON.stringify(args)
+})
+
+export const saToken = googleProject => getServiceAccountToken(googleProject, getUser().token)
```
Comment (jdcanas) [inline src/libs/ajax/GoogleStorage.ts]: consider wrapping this in `GoogleStorage` module to go with the theme of ultimately having all API calls bubble up through Ajax

File: src/libs/ajax/GoogleStorage.ts [hunk 5]
```diff
@@ -1,12 +1,77 @@
...
+      }
+    }
+  }
+  return tryRequest()
+}
+
+// requesterPaysError may be set on responses from requests to the GCS API that are wrapped in withRequesterPays.
+// requesterPaysError is true if the request requires a user project for billing the request to. Such errors
+// are not transient and the request should not be retried.
+export const fetchBuckets = _.flow(withRequesterPays, withRetryOnError(error => Boolean(error.requesterPaysError)), withUrlPrefix('https://storage.googleapis.com/'))(fetchOk)
```
Comment (jdcanas) [inline src/libs/ajax/GoogleStorage.ts]: does this still need to be exported? seems to be a `private` member, which would be nice and ensure all exposed API calls are wrapper in an API-centric module now

Comment (nawatts) [reply src/libs/ajax/GoogleStorage.ts]: Currently, it does not. It may be necessary to export it in the future if we add tests for this module and want to mock this.

File: src/libs/ajax/GoogleStorage.ts (hunk 4 above)
Comment (nawatts) [reply src/libs/ajax/GoogleStorage.ts]: I'm not sure about this one. All the methods of GoogleStorage use the GCS API, whereas this one uses Sam. I could see adding this as a method of Ajax, but I think long term we will probably be moving away from funneling everything through Ajax and using more direct imports of the submodules (GoogleStorage, etc).

File: src/libs/ajax/Billing.js (hunk 3 above)
Comment (nawatts) [reply src/libs/ajax/Billing.js]: Yeah, this is kind of a weird function. I think I'd vote for splitting the `canCompute` part of this out, leaving something like `canUseWorkspaceThroughBillingProject` that always makes a request. I think I'll leave that for later though and keep this PR as reorganization.

File: src/libs/ajax/GoogleStorage.ts (hunk 4 above)
Comment (jdcanas) [reply src/libs/ajax/GoogleStorage.ts]: Yeah, I'm fine with not funneling through Ajax, but I like the idea of all API calls being wrapped in a module where the expectation is that each method makes an api call. Good compromise could be to add a type signature that indicates this returns a Promise.

Comment (nawatts) [reply src/libs/ajax/GoogleStorage.ts]: Good point. Definitely should take advantage of this being in a TS module now to add types.

**Review by cahrens — APPROVED** [review]

File: src/libs/utils.js (hunk 1 above)
Comment (msilva-broad) [reply src/libs/utils.js:94]: I was going to suggest breaking off this accessLevel stuff to a seperate workspace-utils.ts file and use SafeCurry2 pattern to keep the original curried exports intact... thoughts?  A followup PR maybe?  I hesitate because we are in the middle of pushing a common components re-org, and don't wanna take on a utils reorg which feels bigger.

File: src/libs/state.js (hunk 2 above)
Comment (msilva-broad) [reply src/libs/state.js:20]: DRY
It would be better to re-export the getUser export from auth.js here so that the code (even tho it's small) lives in one place and can be changed in one place.

File: src/libs/ajax/GoogleStorage.ts [hunk 6]
```diff
@@ -48,6 +113,24 @@ export type GCSListObjectsResponse = {
 const encodeAnalysisName = name => encodeURIComponent(`notebooks/${name}`)
 
 export const GoogleStorage = (signal?: AbortSignal) => ({
+  checkBucketAccess: async (googleProject, bucket, accessLevel) => {
```
Comment (msilva-broad) [inline src/libs/ajax/GoogleStorage.ts:116]: return types on these methods?

Comment (msilva-broad) [reply src/libs/ajax/GoogleStorage.ts:116]: wait... not seeing much typing at all...

File: src/libs/ajax/Billing.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
+import _ from 'lodash/fp'
+import * as qs from 'qs'
+import { authOpts, fetchBillingProfileManager, fetchOrchestration, fetchRawls, jsonBody } from 'src/libs/ajax/ajax-common'
+
+
+export const Billing = signal => ({
+  listProjects: async () => {
```
Comment (msilva-broad) [inline src/libs/ajax/Billing.js:7]: types on these methods?

File: src/libs/utils.js (hunk 1 above)
Comment (nawatts) [reply src/libs/utils.js:94]: I'd prefer to minimize changes in this PR. I think this style is clearer and easier to understand than the original curried version, so I'm not motivated to try to keep it.

File: src/libs/state.js (hunk 2 above)
Comment (nawatts) [reply src/libs/state.js:20]: Trying to avoid circular imports. auth imports from ajax. But auth could re-export this function from state.

File: src/libs/ajax/GoogleStorage.ts (hunk 6 above)
Comment (nawatts) [reply src/libs/ajax/GoogleStorage.ts:116]: I don't know the return types for these functions. The emphasis of this PR is on reorganization, not functional changes, so I don't think types necessarily need to be added here. Same for the Billing module.

**Review by jdcanas — APPROVED** [review]

**Comment by nawatts** [comment]
@jdcanas was there a separate ticket for this?

**Comment by jdcanas** [comment]
> @jdcanas was there a separate ticket for this?

No, I'll make a subtask under the main one for you to link for compliance, `IA-3832`

**Comment by nawatts** [comment]
Rebased PR branch on dev.
